### DISCUSSION TITLE
Conditional+Looping Timelines

### DISCUSSION THREAD

**Discussion by ritu9901** [jspsych/jsPsych#2409] [Q&A] [answered]
Hello, 
I am trying to create an experiment where the timeline is dependent on the last_trial_data_response. 
```
        var looping_pgs = {
            timeline: [pgs],
            loop_function: function(data) {
                if(jsPsych.pluginAPI.compareKeys(jsPsych.data.getLastTrialData().values()[0].response.name[0],
                'i don't want to answer'){
                return false;
                } else {
                    if(jsPsych.data.getLastTrialData().values()[0].response.name.length===1){
                        console.log('single-parameter-func/page');
                        return true;
                    }
                    else if(jsPsych.data.getLastTrialData().values()[0].response.name.length>1){
                        console.log('multi-parameters-func/page');
                        return true;
                    }
                    else{
                    return false;
                    }
                }    
            }
        }
```
In place of ``` console.log('single-parameter-func/page');``` and ```console.log('multi-parameters-func/page');```, I want to display another trial and as soon as that trial gets over, the loop should continue from where it left.

additional details:
trial ```pgs``` is survey-multi-select
```jsPsych.data.getLastTrialData().values()[0].response.name``` is an array
experiment begins conditionally.

Thanks in advance for any help! I hope I did describe everything precisely enough.
And keep up the very good work! jsPsych is super useful :)

**Comment by becky-gilbert** [comment, marked as answer]
Hi @ritu9901, I think what you need to do is move the logic from the `else` block into a [conditional node](https://www.jspsych.org/7.1/overview/timeline/#conditional-timelines), which will go in your loop timeline after `pgs`. So your looping timeline will be something like `[pgs, new_trial_conditional]`. Then on each iteration of the loop, the `pgs` trial will run, then the conditional node will determine whether the additional trial should run or not before the loop continues.

Also, I don't think you want to use the `compareKeys` function in your if statement - see my comment about that [here](https://github.com/jspsych/jsPsych/discussions/2396#discussioncomment-1801524). 

I hope this helps! Let us know if you need more guidance.

**Reply by ritu9901** [reply]
Hi @becky-gilbert, thank you for getting back to me. I tried what you are suggesting as well, but the problem with this approach is that value of ``` jsPsych.data.getLastTrialData().values()[0].response``` gets changed and hence it creates an error. 
I tried working with ```jsPsych.data.get().filter({exp_name: 'pgs'}).values()```. If there is any way that I could get last response using filter method, I think the problem would be solved.

Also, I really appreciate your tip for not using ```compareKeys``` for such tasks. Thanks :)

**Reply by becky-gilbert** [reply]
Hi @ritu9901, hmm, I'm not sure why `getLastTrialData` doesn't work in the conditional function, because the conditional function should always run after a `pgs` trial. Maybe you could share your code?

In any case, you can use the filter method instead. To get the last trial data you can call the [jsPsych `last` function](https://www.jspsych.org/7.1/reference/jspsych-data/#first-last) on the data collection that you get after filtering, before returning the values.
```js
var last_pgs_trial = jsPsych.data.get().filter({exp_name: 'pgs'}).last(1).values();
```
Does that work?

**Reply by ritu9901** [reply]
heyy @becky-gilbert !
I am sorry for the delay in response. 
```var last_pgs_trial = jsPsych.data.get().filter({exp_name: 'pgs'}).last(1).values();``` worked perfectly!!
Thank you so much.
And, happy new year!! sending love and power :))
